shards/shards.go

```diff
@@ -238,18 +243,32 @@ func selectRepoSet(shards []rankedShard, q query.Q) ([]rankedShard, query.Q) {
 
 	for i, c := range and.Children {
 		var setSize int
-		var hasRepo func(string) bool
+		var hasRepos func([]string) (bool, bool)
 
 		switch setQuery := c.(type) {
 		case *query.RepoSet:
 			setSize = len(setQuery.Set)
-			hasRepo = func(name string) bool {
-				return setQuery.Set[name]
+			hasRepos = func(names []string) (any, all bool) {
+				any = false
+				all = true
+				for _, name := range names {
+					b := setQuery.Set[name]
+					any = any || b
+					all = all && b
+				}
+				return any, all
 			}
 		case *query.RepoBranches:
 			setSize = len(setQuery.Set)
-			hasRepo = func(name string) bool {
-				return len(setQuery.Set[name]) > 0
+			hasRepos = func(names []string) (any, all bool) {
+				any = false
+				all = true
+				for _, name := range names {
+					b := len(setQuery.Set[name]) > 0
```

Comment:
looks like this is the thing that is different in the two hasRepos implementations. Lets pull that out rather